File: C++/2-Medium/00250._统计同值子树/00250._统计同值子树.cpp
#include <iostream>
using namespace std;

// 定义树的节点结构
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// 辅助函数，用于判断是否为同值子树，并统计数量
bool isUnivalTree(TreeNode* node, int& count, int parentVal) {
    if (!node) return true; // 空节点视为同值子树
    
    // 递归检查左右子树
    bool left = isUnivalTree(node->left, count, node->val);
    bool right = isUnivalTree(node->right, count, node->val);
    
    // 如果当前节点不是同值子树，则返回false
    if (!left || !right) return false;
    
    // 如果当前节点是同值子树，增加统计数量
    count++;
    
    // 返回当前节点值是否等于父节点值
    return node->val == parentVal;
}

// 主函数，用于统计同值子树的数量
int countUnivalSubtrees(TreeNode* root) {
    int count = 0;
    // 对于根节点，其父节点值不重要，可以任意设置
    isUnivalTree(root, count, 0);
    return count;
}

// 主函数，用于测试
int main() {
    // 构建测试用例的树
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(1);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(1);
    root->right->right = new TreeNode(1);

    // 计算并输出同值子树的数量
    cout << "The number of unival subtrees is: " << countUnivalSubtrees(root) << endl;

    return 0;
}

File: Python/2-Medium/00250._统计同值子树/00250._统计同值子树.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def countUnivalSubtrees(self, root: Optional[TreeNode]) -> int:
        # 初始化计数器
        self.count = 0
        
        def is_univalue_subtree(node: Optional[TreeNode], parent_value: int) -> bool:
            # 空节点默认为同值子树
            if not node:
                return True
            
            # 递归检查左右子树
            left_check = is_univalue_subtree(node.left, node.val)
            right_check = is_univalue_subtree(node.right, node.val)
            
            # 如果左右子树不是同值子树，返回False
            if not left_check or not right_check:
                return False
            
            # 如果当前节点为同值子树的根，增加计数
            self.count += 1
            
            # 返回当前节点是否为同值子树的一部分
            return node.val == parent_value
        
        # 从根节点开始递归
        is_univalue_subtree(root, -float('inf'))
        return self.count

# 创建测试用例的树结构
root = TreeNode(5, 
                TreeNode(1, TreeNode(5), TreeNode(5)), 
                TreeNode(5, right=TreeNode(5)))

# 实例化Solution并调用函数
sol = Solution()
count = sol.countUnivalSubtrees(root)
print(count)  # 应输出 4，表示找到的同值子树数量